java/memory/src/main/java/org/apache/arrow/memory/LowCostIdentityHasMap.java

```diff
@@ -0,0 +1,363 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one or more
+ *  contributor license agreements.  See the NOTICE file distributed with
+ *  this work for additional information regarding copyright ownership.
+ *  The ASF licenses this file to You under the Apache License, Version 2.0
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.arrow.memory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Highly specialized IdentityHashMap that implements only partial
+ * Map APIs
+ * It incurs low initial cost (just two elements by default)
+ * It assumes Value includes the Key - Implements @ValueWithKeyIncluded iface
+ * that provides "getKey" method
+ * @param <V>
+ */
+public class LowCostIdentityHasMap<V extends ValueWithKeyIncluded> {
+
+  /*
+   * The internal data structure to hold values.
+   */
+  transient Object[] elementData;
```

Comment:
Let's make all the fields private.

Also, no reason to use transient as we aren't using Java serialization.

Reply:
done